runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,264 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+   
+   bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
+
+   // Init class and fields, then jump to end of ICF
+   generateRXInstruction(cg, use64BitClasses ? TR::InstOpCode::STG : TR::InstOpCode::ST, node, classReg, generateS390MemoryReference(targetReg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+   generateSILInstruction(cg, use64BitClasses ? TR::InstOpCode::MVGHI : TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField(), cg), 0);
+   generateSILInstruction(cg, use64BitClasses ? TR::InstOpCode::MVGHI : TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg), 0);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+
+   // We end up in this region of the ICF if the first dimension is non-zero and the second dimension is zero.
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, nonZeroFirstDimLabel);
+
+   TR::Register *componentClassReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, componentClassReg, generateS390MemoryReference(classReg, offsetof(J9ArrayClass, componentType), cg));
+
+   // Calculate maximum allowable object size in elements and ump to OOL if it's higher than firstDimLenReg.
+   int32_t elementSize = comp->useCompressedPointers() ? TR::Compiler->om.sizeofReferenceField() : static_cast<int32_t>(TR::Compiler->om.sizeofReferenceAddress());
+   uintptr_t maxObjectSize = cg->getMaxObjectSizeGuaranteedNotToOverflow();
+   uintptr_t maxObjectSizeInElements = maxObjectSize / elementSize;
+
+   if (comp->target().is64Bit() && !(maxObjectSizeInElements >= 0 && maxObjectSizeInElements <= (uintptr_t)INT_MAX))
+      {
+      uint32_t high32 = maxObjectSizeInElements >> 32;
+      generateRILInstruction(cg, TR::InstOpCode::LLIHF, node, temp1Reg, high32);
+      generateRILInstruction(cg, TR::InstOpCode::IILF, node, temp1Reg, static_cast<uint32_t>(maxObjectSizeInElements));
+      generateRIEInstruction(cg, TR::InstOpCode::CLGRJ, node, firstDimLenReg, temp1Reg, oolJumpLabel, TR::InstOpCode::COND_BRH);
+      }
+   else
+      {
+      generateRILInstruction(cg, TR::InstOpCode::CFI, node, firstDimLenReg, static_cast<uint32_t>(maxObjectSizeInElements));
+      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BHR, node, oolJumpLabel);
+      }
+
+   // Now check to see if we have enough space to do the allocation. If not then jump to OOL code.
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, temp1Reg, firstDimLenReg);
+
+   int32_t round = (elementSize >= TR::Compiler->om.objectAlignmentInBytes()) ? 0 : TR::Compiler->om.objectAlignmentInBytes();
+   int32_t disp32 = round ? (round - 1) : 0;
```

Comment:
Is this ever even true? I don't think field width is ever smaller than object alignment.

Reply:
I don't think it's possible. Object alignment is computed in [1] and it returns the maximum of `OMR_MINIMUM_OBJECT_ALIGNMENT` which is 8 [2], and the other value. So it is always at least 8. There are no array elements that are greater than 8 bytes in size, so this entire computation can be simplified.

[1] https://github.com/eclipse/omr/blob/94b8b73a050f72c6afcfdfeafe94c897b3892794/gc/base/ObjectModelBase.hpp#L336
[2] https://github.com/eclipse/omr/blob/94b8b73a050f72c6afcfdfeafe94c897b3892794/gc/base/ObjectModelBase.hpp#L50

Reply:
I was currently looking into this. But I'll try your suggestion first.

Reply:
Thought about this again. So `elementSize` is only 4 or 8 bytes realistically, so under 64-bit compressedrefs we can have the false case of the ternary here. We should use the `OMR::align` API here though.

```
int32_t elementSizeAligned = OMR::align(elementSize, TR::Compiler->om.objectAlignmentInBytes());
int32_t alignmentCompensation = (elementSize == elementSizeAligned) ? 0 : elementSizeAligned - 1;

...

if (alignmentCompensation != 0)
   {
   generateRILInstruction(cg, TR::InstOpCode::NILF, node, temp1Reg, -alignmentCompensation);
   }
```

Reply:
Using the correct API in 0c0a523.